crates/types/src/invocation.rs

```diff
@@ -273,19 +273,95 @@ impl fmt::Display for InvocationTarget {
     }
 }
 
+/// Concurrency guarantee of the invocation request.
+#[derive(Debug, Default, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+enum ConcurrencyGuarantee {
+    /// Enqueue the invocation in the inbox when the target is busy
+    EnqueueWhenBusy {
+        /// fka ServiceId.name
+        inbox_target: ByteString,
+        /// fka ServiceId.key
+        inbox_key: ByteString,
+    },
+    /// No queueing, just execute the request
+    MaxParallelism,
+    /// Use the default from the target semantics
+    #[default]
+    UseTargetDefault,
+}
+
+impl ConcurrencyGuarantee {
+    fn infer_target_default(invocation_target: &InvocationTarget) -> ConcurrencyGuarantee {
+        match invocation_target {
+            InvocationTarget::VirtualObject { handler_ty: VirtualObjectHandlerType::Exclusive, name, key, .. } => {
+                ConcurrencyGuarantee::EnqueueWhenBusy {
+                    inbox_target: name.clone(),
+                    inbox_key: key.clone(),
+                }
+            }
+            InvocationTarget::Service { .. } |
+            InvocationTarget::VirtualObject { handler_ty: VirtualObjectHandlerType::Shared, .. } |
+            /* For workflow, there is no enqueueing as we have the behavior on existing invocation id that guarantees correctness */
+            InvocationTarget::Workflow { .. }  => ConcurrencyGuarantee::MaxParallelism
+        }
+    }
+}
+
+/// Behavior when sending an invocation request and the request already exists.
+#[derive(Debug, Default, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+enum BehaviorOnExistingInvocationId {
+    /// Attach to the existing invocation
+    Attach,
+    /// Reply with "conflict" error response
+    ReplyConflict,
+    /// Just drop the request
+    Drop,
+    /// Use the default from the target/idempotency key semantics
+    #[default]
+    UseTargetDefault,
```

Comment:
I'm not sure I like having the default variant. It adds cognitive overhead. It sounds like "it depends". An alternative strategy is to always set it on the caller-side, I suppose.